File: src/Composer/Project/Assets.cpp
/**************************************************************************

 The MIT License (MIT)

 Copyright (c) 2015 Dmitry Sovetov

 https://github.com/dmsovetov

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:
 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.
 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 **************************************************************************/

#include "Assets.h"
#include "Project.h"

#include "../Importers/ImageImporter.h"
#include "../Importers/FileImporter.h"
#include "../Importers/MeshImporter.h"

DC_BEGIN_COMPOSER

// ** Assets::Assets
Assets::Assets( QObject* parent, const Io::Path& path, AssetFileSystemModelQPtr assetFileSystem ) : QObject( parent ), m_path( path ), m_assetFileSystem( assetFileSystem )
{
	// Create an asset bundle
	m_bundle = Scene::AssetBundle::create( "Assets", path );

	// Declare asset importers.
#ifdef HAVE_TIFF
	m_assetImporters.declare<Importers::ImageImporterTIF>( "tif" );
#endif	/*	HAVE_TIFF	*/

#ifdef HAVE_FBX
	m_assetImporters.declare<Importers::MeshImporterFBX>( "fbx" );
#endif	/*	HAVE_FBX	*/

	m_assetImporters.declare<Importers::ImageImporterTGA>( "tga" );
	m_assetImporters.declare<Importers::FileImporter>( "material" );

	// Connect to asset model signals
    connect( m_assetFileSystem, SIGNAL(fileAdded(const FileInfo&)), this, SLOT(addAssetFile(const FileInfo&)) );
    connect( m_assetFileSystem, SIGNAL(fileRemoved(const QString&, const FileInfo&)), this, SLOT(removeFromCache(const QString&, const FileInfo&)) );
    connect( m_assetFileSystem, SIGNAL(fileChanged(const QString&, const FileInfo&)), this, SLOT(updateAssetCache(const QString&, const FileInfo&)) );
}

// ** Assets::bundle
Scene::AssetBundleWPtr Assets::bundle( void ) const
{
	return m_bundle;
}

// ** Assets::createAssetForFile
Scene::AssetPtr Assets::createAssetForFile( const FileInfo& fileInfo )
{
	// Get the asset type by extension
	Scene::Asset::Type type = assetTypeFromExtension( fileInfo.extension() );

	// Create asset by type
	Scene::AssetPtr asset = m_bundle->createAssetByType( type );

	if( !asset.valid() ) {
		return Scene::AssetPtr();
	}

	// Set meta file
	m_assetFileSystem->setMetaData( fileInfo, asset->keyValue() );

	return asset;
}

// ** Assets::registerExtension
void Assets::registerExtension( const String& extension, Scene::Asset::Type type )
{
	m_assetTypes[extension] = type;
}

// ** Assets::assetTypeFromExtension
Scene::Asset::Type Assets::assetTypeFromExtension( const String& extension ) const
{
	AssetTypes::const_iterator i = m_assetTypes.find( extension );
	return i != m_assetTypes.end() ? i->second : Scene::Asset::TotalAssetTypes;
}

// ** Assets::removeAssetFromCache
void Assets::removeAssetFromCache( const QString& uuid, const FileInfo& file )
{
	// Remove asset from bundle
	m_bundle->removeAsset( uuid.toStdString() );

	// Remove asset from cache
    qComposer->fileSystem()->removeFile( cacheFileFromUuid( uuid.toStdString() ).c_str() );
}

// ** Assets::updateAssetCache
bool Assets::updateAssetCache( const QString& uuid, const FileInfo& file )
{
	// Create an asset importer for
	Importers::AssetImporterPtr importer = m_assetImporters.construct( file.extension() );

	if( !importer.valid() ) {
		return false;
	}

    // Get the shared file system instance
    FileSystemQPtr fs = qComposer->fileSystem();

	// Create the Assets folder
	Io::Path AssetsFolderPath = cacheFolderFromUuid( uuid.toStdString() );

	if( !fs->fileExists( AssetsFolderPath.c_str() ) ) {
		fs->createDirectory( AssetsFolderPath.c_str() );
	}

	// Get Assets file path
	Io::Path AssetsFilePath = cacheFileFromUuid( uuid.toStdString() );

	// Don't Assets twice
	if( fs->fileExists( AssetsFilePath.c_str() ) ) {
		return true;
	}

	// Perform asset caching.
	bool result = importer->import( fs, file.absolutePath(), AssetsFilePath );
	DC_BREAK_IF( !result );

	qDebug() << "Cached" << file.fileName().c_str();

	return result;
}

// ** Assets::addAssetFile
void Assets::addAssetFile( const FileInfo& fileInfo )
{
	// Read the meta data
	Archive meta = m_assetFileSystem->metaData( fileInfo );

	// Added asset
	Scene::AssetPtr asset;

	// Create asset from data or create the new one
#if DEV_DEPRECATED_KEYVALUE_TYPE
	if( !meta.isNull() ) {
#else
    if( meta.isValid() ) {
#endif  /*  DEV_DEPRECATED_KEYVALUE_TYPE    */
        asset = m_bundle->createAssetFromData( meta );
	} else {
		asset = createAssetForFile( fileInfo );
	}

	if( !asset.valid() ) {
		return;
	}

	// Set asset name
	String name = fileInfo.relativePath( m_assetFileSystem->rootPath() );
	bool result = m_bundle->setAssetName( asset, name );
	DC_BREAK_IF( !result );

	// Add asset to bundle
	qDebug() << "Added" << asset->name().c_str();
	m_bundle->addAsset( asset );

	// Put asset to cache
	updateAssetCache( QString::fromStdString( asset->uuid() ), fileInfo );
}

// ** Assets::cacheFolderFromUuid
Io::Path Assets::cacheFileFromUuid( const String& uuid ) const
{
	return cacheFolderFromUuid( uuid ) + uuid;
}

// ** Assets::cacheFolderFromUuid
Io::Path Assets::cacheFolderFromUuid( const String& uuid ) const
{
	String folder = String() + uuid[0] + uuid[1];
	return m_path + folder;
}

DC_END_COMPOSER

File: Source/CLI/makedata/assets/assets.py
#################################################################################
#
# The MIT License (MIT)
#
# Copyright (c) 2015 Dmitry Sovetov
#
# https://github.com/dmsovetov
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.
#
#################################################################################

import os
import fnmatch
import random
from file import File


class Assets(object):
    """Loaded assets"""

    def __init__(self, input_path, output_path, cache, build_rules, importers=None):
        """Constructs assets instance that are loaded from a specified path"""
        self._input_path = input_path
        self._output_path = output_path
        self._files = {}
        self._outdated = []
        self._build_rules = build_rules
        self._cache = cache
        self._importers = importers

    @property
    def outdated(self):
        """Returns a list of outdated assets"""
        return self._outdated

    @property
    def files(self):
        """Returns all known asset files"""
        return self._files

    @property
    def input_path(self):
        """Returns an input assets path"""
        return self._input_path

    @property
    def output_path(self):
        """Returns an output assets path"""
        return self._output_path

    @property
    def cache(self):
        """Returns a cache instance"""
        return self._cache

    @property
    def build_rules(self):
        """Returns build rules"""
        return self._build_rules

    @property
    def importers(self):
        """Returns available asset importers"""
        return self._importers

    @classmethod
    def generate_uuid(cls):
        return '%032x' % random.getrandbits(128)

    def importer_by_file_name(self, file_name):
        """Returns a default asset importer by a file name"""

        for k, importer in self.build_rules:
            if fnmatch.fnmatch(file_name, k):
                return importer()

        return None

    def _add_outdated_asset(self, asset_file):
        """Puts an asset file to an outdated list"""
        self._outdated.append(asset_file)

    def _add_asset(self, asset_file):
        """Puts an asset file to a registry"""

        if asset_file.uuid in self.files.keys():
            raise ValueError("duplicate asset id '%s'" % asset_file.uuid)

    def _create_asset(self, local_path):
        """Creates an asset from file"""

        absolute_input_path = os.path.join(self.input_path, local_path)

        importer = self.importer_by_file_name(local_path)

        if not importer:
            return None

        asset = File(absolute_input_path, uuid=self.generate_uuid(), importer=importer)
        asset.save()

        return asset

    def scan(self):
        """Scans an input directory for asset changes"""

        print 'Scanning %s...' % self.input_path

        for root, dirs, files in os.walk(self.input_path):
            for file_name in files:
                name = os.path.splitext(file_name)[0].lower()
                input_path = os.path.join(root, file_name)
                relative_path = os.path.relpath(input_path, self.input_path)

                if name.startswith('.'):
                    continue

                if file_name.endswith(File.META_EXT):
                    continue

                asset = File.load(input_path, self.importers)

                if asset is None:
                    asset = self._create_asset(relative_path)

                if not asset:
                    continue

                if self.cache.update_asset_hash(asset.uuid, meta_hash=asset.meta_hash, file_hash=asset.file_hash):
                    self.outdated.append(asset)
                elif not os.path.exists(os.path.join(self.output_path, asset.local_output_path)):
                    self.outdated.append(asset)

    def build(self, task_manager):
        """Builds each outdated file to an output folder"""

        def make_importer(asset_path, asset):
            return lambda: {importer.process(asset_path, asset) for name, importer in asset.importers.items()}

        for item in self.outdated:
            task_manager.push(make_importer(self.output_path, item))

